### DISCUSSION TITLE
Welcome to CiliaQ Discussions!

### DISCUSSION THREAD

**Discussion by hansenjn** [hansenjn/CiliaQ#7] [General]
# ![CiliaQ](https://github.com/hansenjn/CiliaQ/blob/master/Webfiles/20200618%20CiliaQ%20Logo%20Small.png?raw=true) Discussions
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of the CiliaQ community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together.

**Comment by aaron174** [comment]
Hello!   Are there test .tiff files available to start learning how to use CiliaQ?  We are just starting cells for our cilia assay in the lab and I was hoping to get started early on image analysis.  
Best regards, 

Aaron

**Reply by hansenjn** [reply]
Hello Aaron,

Thank you for your question and, at the same time, suggestion. We now provide an [examples folder in the github repository](https://github.com/hansenjn/CiliaQ/tree/master/Examples) with a first example image. We will add more examples soon. In addition, we are also right now planning and producing tutorials in written and video form right now, which will be available online as soon as possible.

Best regards
Jan

**Reply by aaron174** [reply]
Hi Jan, 
Thank you for providing the example folder and test image.  I was excited to see your paper and software available.  It's very good timing as I'm looking to develop an assay for cilia expression.  
Best regards, 
Aaron

**Comment by lbozal** [comment]
Hello!! Thanks for your hard work!! It is very helpful!
Would it be possible to implement a new measurement parameter: cilia percentage or number of cilia vs number of cells? In order to do so, CiliaQ should be able to detect nuclei... maybe the user can let it know that one of the channels is nuclei (DAPI).
Best,

Laura

**Reply by hansenjn** [reply]
Hello Laura,
thank you very much for that suggestion - that's a very good idea and that's possible. I will put it on the TODO list and will try to implement it in the next CiliaQ version.
Best wishes
Jan

**Reply by cruivinho** [reply]
Hello, were you already able to implement this addition?

**Reply by hansenjn** [reply]
Hi @cruivinho,

Sorry that this took a while. The short answer: you can already do this today, and a more convenient implementation is still in the works.

Since CiliaQ v0.2.1 you can define one channel as the basal body channel and set it to "segmented - detect and quantify BBs from channel". CiliaQ then detects the objects in that channel and assigns each of them to the closest cilium. While if you can include a PCNT stain to segment and use basal bodies can be the most straightforward, the channel does not need to contain basal bodies - you can equally put a (segmented) nuclei staining in there, and in turn you obtain the number of cells, how many of them carry a cilium, and thus the percentage of ciliated cells.

This is exactly how we did it in our recent paper: https://www.cell.com/cell/fulltext/S0092-8674(25)01029-3

A few practical points:

- The nuclei channel needs to be segmented with CiliaQ Preparator or something else first, just like the reconstruction channel.
- In the basal body detection dialog you can set the minimum object size, whether to increase the range when connecting pixels to an object, and the maximum distance allowed between an object and the end of a ciliary skeleton. For nuclei, this distance needs to be set considerably larger than for real basal bodies, since the center of a nucleus lies further away from the cilium.
- You can also choose whether objects without cilia and cilia without objects are kept or excluded from the results.

v0.2.1 is not yet distributed via the FIJI update site since it does not yet fully implement this option for timelapse images and has been only tested extensively for static (1 time frame) images, so please download the .jar from the release page and install it manually: https://github.com/hansenjn/CiliaQ/releases/tag/v0.2.1.

How to? https://github.com/hansenjn/CiliaQ/wiki/Q&A:-How-to-install-a-specific-version-of-a-CiliaQ-plugin%3F

As also emphasized above, there are two limitations to be aware of:
- This does not yet work for time-lapse images - in 4D the analysis is cancelled. Frame-to-frame tracking is currently being implemented and will come with v0.2.2.
- I am still working on making this more convenient and properly documented, which will probably take a few more weeks to months. Until then, the basal body option above should already give you what you need, I hope.
- The v0.2.1 version will throw you also a warning that this is a version in testing, but you can use the "static nontimelapse" version, I tested it extensively. Still please let me know if something looks unusual or have a look whether the values you will get makes sense (e.g., a cilium that seems short is short, a cilium that seems long is long).
- More information is in the release notes also for the last releases (v0.2.0 and v0.2.1: https://github.com/hansenjn/CiliaQ/releases/).

Feedback on how it performs on your data is very welcome. And please ask more questions if you have some, will make sure to answer faster now.

Best wishes
Jan

**Comment by drwty** [comment]
Hello! I am a complete beginner to CilliaQ plugin. I have downloaded the test image and analyzed by the CiliaQ. After imported the image by imageJ and launched the preparator plugin, I used the default settings to segment the red channel. However, the output _CQP.tif file showed no change as compared with original file. In the example image provided in the website named _CQP.tif, red channel staining the primary cilia was changed to white color. I think some settings might be wrong during analysis. Could you tell me the reason? Thank you very much. 

Best,

Drwty

**Reply by hansenjn** [reply]
Dear Drwty,

thank you for trying out CiliaQ and informing us that the information on the example data set was not complete.

In the CQP file available in the repository I had manually changed the color of the third (segmented) channel to cyan after the file was output by CiliaQ - in turn the cilia appeared in white. This is not necessary to be done (and I had forgotten that I did this for the example data set). However, it can help when inspecting the CQP.tif file. To clarify why I did this and how exactly we processed the example data set (we also did not use the default settings by CiliaQ Preparator), I extended the instructions in the Readme file for the example data set (see here: https://github.com/hansenjn/CiliaQ/blob/master/Examples/MEF%20SAG-treated/).

Generally, when you compare the CQP.tif file and the original file, both files may look identical on the first glance. However, when you look at the number of channels you will see a difference. CiliaQ Preparator (by default) duplicates the channel to be segmented. It segments only a copy of the channel. In the segmented channel, background pixels are set to zero and intensity in the foreground pixels is preserved. Thus the overlay of all channels will look identical for the CQP.tif and the original file. To clearly see how the segmentation performed you may need to switch to the segmented channel using the "C" bar on the bottom of the image in FIJI / ImageJ, eventually change its false-color, and adjust the brightness / contrast of the image display (how this is done is explained in the updated readme file now).

Hope this helps - Feel free to ask if something is still unclear.

Best

Jan

**Reply by drwty** [reply]
Hi, Jan

Thanks for your kindly reply. I have tried again following the instructions written in the readme file and it works well. 
I will start to analyze my cilia staining on brain slices. If there is any question I will contact with you. 
Thank you again.

Best,

Drwty

**Comment by Harry-D** [comment]
Hi Jan, 
Thank you for the awesome CiliaQ plug-in you put together. I just installed it and it worked very well with the sample image you provided. Somehow when I use my own image (...CQP.tif <- processed with CQPreparator that includes the segmented arl13b channel; please refer to the attached images), right after the status "copying result to output file", Fiji started a thinning iteration (image also attached), that does not end even after 130 iterations... I don't think thinning iteration is included in your SOP, and for general image thinning, 5 iterations are usually enough. Could you help me identify what I did wrong here? Thank you for your help!

Best,
Harry

I have attached some screenshots of the different channels and the thinning iterations
<img width="435" alt="Screen Shot 2021-05-20 at 2 47 03 PM" src="https://user-images.githubusercontent.com/54862193/119032585-59e10580-b97a-11eb-95df-8887b7d5fc41.png">
<img width="417" alt="Screen Shot 2021-05-20 at 2 46 53 PM" src="https://user-images.githubusercontent.com/54862193/119032588-5a799c00-b97a-11eb-9b64-742398bf7e2a.png">
<img width="487" alt="Screen Shot 2021-05-20 at 2 47 12 PM" src="https://user-images.githubusercontent.com/54862193/119032581-577eab80-b97a-11eb-94ef-2efe992c7ac0.png">
<img width="610" alt="Screen Shot 2021-05-20 at 2 51 01 PM" src="https://user-images.githubusercontent.com/54862193/119033104-d4aa2080-b97a-11eb-8b2c-cfed68a539f3.png">

**Reply by hansenjn** [reply]
Hi Harry,

I am happy to help!

My first guess is that you might have specified an incorrect channel as the "reconstruction" channel in CiliaQ. When a channel that is not segmented is specified as the "reconstruction" channel, CiliaQ may run for a long time and will need to perform a lot of thinning iterations because it will detect the whole image as one big particle. In the example data set, we wanted to detect cilia based on the third channel. Thus, we specified in CiliaQ Preparator the channel 3 as the channel to be segmented. In turn, in the output CQP.tif image, channel 3 was segmented (non-ciliary pixels set to zero) and we had to specify channel 3 as the reconstruction channel in CiliaQ, from which cilia masks then were detected by CiliaQ. 

In your analysis, the first channel contains the ciliary labeling and thus, you correctly segmented the first channel via CiliaQ Preparator. Accordingly, you would need to define channel 1 as the reconstruction channel:
![Screenshot](https://user-images.githubusercontent.com/27991883/119197838-ac104c80-ba88-11eb-822e-66d076d77cef.png)
Did you do this? In case you correctly set this, please let me know and I will think about what else could have caused the extensive thinning iterations.

Best wishes
Jan

**Reply by Harry-D** [reply]
Hi Jan, 
Thank you for the prompt reply and thank you for point out the mistake. Somehow in the CQP.tif file, the coloring was lost and I erroneously labeled the unsegmented channel as cyan. I guess for future reference, one can always refer to the CQP.txt file to know for sure which channel is the newly created/segmented channel. 

For Laura's previous question on cell counting. I'm actually doing the exact same thing using other plug-ins. It would make my life so much easier If you could combine this feature into CiliaQ, and output cell number/percentage ciliated cells. Thanks again!

Best,
Harry

**Comment by ecenazbilgic** [comment]
Hi! Thank you for this useful plugin! I am currently analyzing my data for ARL13B staining and am interested in the output for the cilia length.  I have come across so far that around 20 percent of the cilia length for the found cilia is either 0 or NA. In one of the explanations, I saw that when the cilia is spherical, the length column is left empty but then I was wondering what sort of cilia shape gives a result of 0 length. How should I deal with these cases? Do you suggest that I should filter those results out or is there anything else that I can do?

Thank you for your time and help!

Ece

**Reply by hansenjn** [reply]
Hi Ece (@ecenazbilgic),
yes, it is as you say that cilia whose detected shape is spherical will emerge as a length value of zero.

I will try to explain why you get spherical objects and why the length of them is detected as zero / NA.

In CiliaQ, the length of the cilium is determined based on the skeleton of the ciliary "mask". This is valid since cilia have a width that lies at the diffraction limit - thus, there appearance of a cilium in light-based microscopy images appears "blurred", even at a decent "scanning" rate / camera where you reach pixel widths below the wavelength of light (e.g., 0.05 µm). Consequentially, the actual length of the cilium corresponds only to the central line of a ciliary "appearance"/ segmentation mask:
![image](https://user-images.githubusercontent.com/27991883/211896521-50ddba45-9c66-4d64-aa43-0391f0e7905e.png)

Deriving the skeleton and using that as the length measurement has also a second advantage. It makes the analysis less sensitive to the segmentation method selected by a user. For different segmentation methods you may detect slidely wider or narrower ciliary masks (e.g. wide masks with Canny3D since it includes the edges of the appearing cilium, while narrower masks with a harsh threshold algorithm like "RenyiEntropy") - however, the skeleton is the almost the same even when applying it to images generated with different segmentation methods. I tried to draw it here:
![image](https://user-images.githubusercontent.com/27991883/211897565-90d3c824-63d0-42fd-a9cb-d4fed8989901.png)

If one would go with the length measurements from begin of a mask to end of a mask the measurement would be very sensitive to different segmentation methods and easily biased by the segmentation method.
![image](https://user-images.githubusercontent.com/27991883/211900763-06aa49e4-dd7b-44fb-9584-a8013d405d14.png)

Now if we see a "mask" for a cilium that is like a sphere the skeleton will be a point and thus has no length:
![image](https://user-images.githubusercontent.com/27991883/211897788-b9c463fc-b4f2-4f81-b4f9-be98529f14a8.png)

Of course "in real" this cilium might have a length. Such "sphere" cilia structures in the image may relate to:
- In the case of analyzing a 2D image in CiliaQ (no different z planes, e.g., acquiring an epifluorescent image or only one confocal plane, or analyzing a maximum intensity projection of a stack): the sphere cilium may be only a cross-section through the actual cilium or it may relate (in non-confocal microscopy) to a "standing" cilium (a cilium aligning with the imaging axis = the z axis = the objective). 
- In case you analyze a 3D image: It could be that it relates to a cilium at the edge of the stack and thus you see only a fraction of it - or it could be a ciliary "stub" (e.g. emerging from a degrading, shedded of, or forming cilium, which your marker of choice labels only as a "point" that appears as a spherical mask in CiliaQ after segmentation).

In any case I would include these cilia into your quantification. What I usualy did is that I either include them with a length of zero or that I just added to all cilia in my experiment (those that give length values and those that give zero length values) a fixed value corresponding to one pixel length or less (e.g., 0.1 µm). It is justifiable since the "spherical" zero-length may actually refer to a ciliary length below the diffraction limit. If you do it that way, one does not have zero values for a length and if one adds the same value to all length measurement in the same experiment this does not manipulate results / conclusions.

Hope this helps you and my explanations are clear - feel free to ask more questions / let me know if some things are not clear.

Best wishes
Jan

**Reply by ecenazbilgic** [reply]
Hi Jan,

Thank you so much for such a fast and detailed explanation! Now it is really clear for me how the length calculation works and what my results indicate. I will keep these explanations in mind for my analysis.

Best,
Ece

**Comment by cruivinho** [comment]
Hello,

I have a question about length measurement for cilia oriented perpendicular to the imaging plane.

In my data (confocal, 0.1833 µm lateral pixel size, 0.240 µm z-step), a substantial fraction of cilia extend mostly along Z: they span less than 1 µm in XY but 2–4 µm in Z, appearing as puncta in any single XY slice. Will CiliaQ measure their true 3D length, or is there a risk that such objects are skeletonized into a single point and reported with near-zero length?
See image as example. 
<img width="1421" height="760" alt="pontos_sao_cilios_verticais" src="https://github.com/user-attachments/assets/2c15d62a-232c-40fb-8fc1-430b46fa9b00" />

Thank you

**Reply by hansenjn** [reply]
Hi @cruivinho,

Thank you for the question - and for adding the example image, that helps!

Short answer: no, these cilia will not be reduced to a point. Skeletonization in CiliaQ works in 3D, so an object that is small in XY but extends over many planes in Z is still an elongated object and will give a proper skeleton. The "spherical" case I described further up in this thread applies only to objects that are compact in all three dimensions, not to cilia standing along the imaging axis.

The length is calculated in calibrated 3D as well - the distance between skeleton points includes the z component multiplied by the voxel depth. With your 0.240 µm z-step, a cilium of 2-4 µm spans roughly 8 to 17 planes, which is plenty for a reliable skeleton.

Two things I would still check, because they affect exactly these cilia:

1. Calibration. Go to Image > Properties and confirm that the voxel depth really reads 0.240 µm. For a cilium standing in Z the length depends almost entirely on the z calibration, so a wrong or missing z value shows up in this population first and nowhere else. If not, you can also define the z calibration in CiliaQ itself.

2. Border exclusion. A cilium running along Z is much more likely to touch the first or last plane of the stack, and by default CiliaQ excludes such objects. So the risk in your data is not a near-zero length but the cilium disappearing from the results completely - it then appears with a grey ID in the ..._CQ_RP_3D.tif file. I would record a few empty planes above and below the cells and check the 3D output for grey IDs. See also: https://github.com/hansenjn/CiliaQ/wiki/Q&A:-After-CiliaQ,-results-for-cilia-labeled-grey-in-the-3D-output-images-are-not-output.-Why%3F Or otherwise, you can also just seelct CiliaQ to only exclude x and y touching cilia.

To verify on the results you already have: in the ..._CQ.txt file, look at the cilia length to validate this.

One caveat to keep in mind: the axial resolution of a confocal is 2-3x worse than the lateral one, so masks of cilia standing in Z are stretched a little along the imaging axis and their length may be slightly overestimated compared to cilia lying in the imaging plane. If you compare groups, it is worth checking whether the orientation vector z relates to your length values.

Also certainly you always need to get a good segmentation strategy going.

I am currently working on improved versions for CiliaQ, CiliaQ Preparator, and the wiki, and will also have some updates that make the workflow easier soon. CiliaQ Preparator v0.2.0 is already released but not yet on the FIJI update page - will also follow soon. CiliaQ Preparator v0.2.0 now also allows to run it from Macros or headless (https://github.com/hansenjn/CiliaQ_Preparator/releases).

Feel free to ask more questions or send more screenshots if numbers are off or something seems to not work - happy to have a look and always happy for feedback.

Thanks for using CiliaQ!

Best wishes
Jan
